aurt/recursive_estimator.py

```diff
@@ -0,0 +1,58 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from logging import Logger
+
+from aurt.caching import Cache
+from aurt.linear_system import LinearSystem
+
+
+class RecursiveEstimator(ABC):
+    """
+    An abstract base class for an estimator for a linear system. The method requiring implementation are:\n
+      - gain_matrix()
+    """
+    
+    @abstractmethod
+    def __init__(self, logger: Logger, cache: Cache, linear_system: LinearSystem, name: str=None, parameters_init=None):
+        self.logger = logger
+        self._cache = cache
+        self._name = name
+        self._linear_system = linear_system
+
+        if parameters_init is not None:
+            self.parameters = parameters_init
+        else:
+            self.logger.info(f"No initial values for parameter estimates were provided for '{self.name}'. Assuming zero for all initial values.")
+            self.parameters = np.zeros((sum(linear_system.number_of_parameters()), 1))
+    
+    @property
+    def name(self) -> str:
+        return self._name
+    
+    @name.setter
+    def name(self, name: str):
+        self._name = name
+
+    @property
+    def parameters(self) -> np.array:
+        """
+        An 'n_joints'-element list with each element 'j' comprising a list of base parameters related to joint 'j'.
+        """
+        return self._parameters
+    
+    @parameters.setter
+    def parameters(self, parameters: np.array):
+        self._parameters = parameters
+    
+    def update_parameters(self, measurement: np.array, states: np.array) -> np.array:
+        observation_matrix = self._linear_system.observation_matrix(states)
+        eps = measurement - observation_matrix @ self.parameters  # Innovations
+        self.parameters = self.parameters + self.gain_matrix(observation_matrix) @ eps  # Parameter estimate
+        return self.parameters
+    
+    @abstractmethod
+    def gain_matrix(self, observation_matrix: np.array) -> np.array:

```

Comment:
Shouldn't this return np.ndarray like in linear_system.py line 220 ? Or what is the difference between using np.array and np.ndarray?

Reply:
I'm a retard, recursive_estimator.py should be deleted from this branch because it's part of the work on online estimation

Reply:
...but yes, you are correct, np.array is a method and np.ndarray is the data type that would have to be returned here